### PR TITLE
add utils and decomposition

### PR DESCRIPTION
**Pull request by ChenZhao44**
`tcount, random_circuit, decompose_zx`

### CODE REVIEW

File: src/decomposition.jl [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    end
+end
+function decompose_zx(cg::Ctrl)
+    ctrl = cg.ctrl
+    g = cg.gate
+    if length(ctrl) == 1
+        if length(g.locations) == 1
+            @switch g begin
+                @case Gate(::YGate, loc)
+                    return Chain(Gate(YaoHIR.S', loc), Ctrl(Gate(YaoHIR.X, loc), ctrl), Gate(YaoHIR.S, loc))
```
Comment (Roger-luo) [inline src/decomposition.jl:44]: no need to write the module name inside YaoHIR module

File: src/utils.jl [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+end
+
+"""
+    tcount(c)
+
+Returns the number of non-Clifford gates in circuit `c`. 
+"""
+tcount(bir::BlockIR) = tcount(bir.circuit)
+tcount(c::Chain) = sum(tcount(g) for g in c.args)
+tcount(::T) where {T <: Routine} = error("T-count for type $T is not defined")
```
Comment (Roger-luo) [inline src/utils.jl]:
```suggestion
tcount(::Routine) = error("T-count for type $T is not defined")
```

File: src/utils.jl [hunk 3]
```diff
@@ -0,0 +1,61 @@
+using CompilerPluginTools, YaoLocations
+
+"""
+    random_circuit(nbits, ngates [, two_qubits_ratio, t_ratio])
+
+Generate a random quantum circuit of qubit number `nbits` and with gate number `ngates`,
+where 2-qubit gate number is of ratio `two_qubits_ratio` and T-gate number of ratio `t_ratio`.
+"""
+function random_circuit(nbits, ngates, two_qubits_ratio = 0.2, t_ratio = 0.1)
```
Comment (Roger-luo) [inline src/utils.jl]: we should implement this directly as a normal `@device` function, this is over complicated.

Comment (Roger-luo) [reply src/utils.jl]: for validation, we will just directly evaluate things in YaoLang instead of constructing IR.

Comment (Roger-luo) [reply src/utils.jl]: besides, this implementation won't compile... you didn't construct the correct `IRCode`, let just remove this function in this PR for now, I'll do it in YaoCompiler.

**Review by Roger-luo — COMMENTED** [review]
I'm still not sure if decomposition passes should be in YaoCompiler, I think they should. But let's fix the small issues I commented above and remove `random_circuit`, then this PR should be good to go.

File: src/decomposition.jl (hunk 1 above)
Comment (ChenZhao44) [reply src/decomposition.jl:44]: It will get errors if there is no YaoHIR. Maybe it is an issue of `@switch` and `@case`.

**Comment by Roger-luo** [comment]
@ChenZhao44 need to have some more tests for the decomposition part: https://app.codecov.io/gh/QuantumBFS/YaoHIR.jl/compare/5/tree/src/decomposition.jl

**Comment by ChenZhao44** [comment]
I will add some.

File: src/utils.jl [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+function tcount(g::Ctrl)
+    if length(g.ctrl) == 1 && length(g.gate.locations) == 1
+        (g.gate.operation === X || g.gate.operation === Z) && return 0
+    end
+    return tcount(decompose_zx(g))
+end
+tcount(g::Gate) = tcount(g.operation) * length(g.locations)
+tcount(::IntrinsicRoutine) = 0
+tcount(::TGate) = 1
+tcount(g::Rx) = (abs(rem2pi(g.θ*4, RoundNearest)) < 1e-10 ? 0 : 1)
```
Comment (Roger-luo) [inline src/utils.jl]: I think the 1e-10 bound Jinguo uses was incorrect, the floating-point precision should be `sqrt(eps(Float64))` which is around `1e-8`

Comment (Roger-luo) [reply src/utils.jl]: we need to check this based on the type of theta

**Pull request merged**